Update HEOM for v5.

hodgestar opened the pull request:
**Description**
Update HEOM for v5.

Other changes:
- This resulted in a bit of a refactor of the Result class (#1907)
- I found a few places where QobjEvo did not correctly set `.type` and fixed those (they were breaking the HEOM tests).

Todos and questions:
- [X] Should HierarchyADOsState store ado_state as a Data object rather than a numpy array? (it can't be stored as a Data object because the ADOs have shape `n_ado, n, n` and the Data layer only supports 2-d arrays).
- [X] How should the ado_state be stored on a results object? In particular, how should the change in e_ops signature be handled (for the HEOM solver the e_ops callables also take the full ADO state). (implemented)
- [X] What should be done with the steady_state method? I guess it can just stay. Should other solvers also have this? (it should just stay, I think)
- [x] Recheck the time-dependent case. (I also simplified the time-dependent case so that the Liouvillian is never added and so doesn't have to be subtracted off).
- [X] The HEOM can't really take new QObjEvo arguments each time `run` is called because calculating the RHS again is quite expensive. Maybe something can be done? (implemented -- passing new args to .run is now supported)
- [X] In the base solver the new options seem to be set after retrieving the integrator -- is this right? (fixed)
- [X] Compare performance to the HEOM in 4.7. (looks faster, yay!)
- [X] We could also implement `.start()` and `.step()` for the HEOMSolver. Not quite sure of the use case yet though. (implemented and tested -- the implementation came for free after sorting out having .run inherit from the base solver).
- [x] Add some direct tests for the QobjEvo type errors that have been fixed.
- [x] Moved solver into `qutip.solver.heom`. 

**Related issues or PRs**
- One of the todos from #1850
- In #1907 the Result class was changed largely to support extension by the HEOMSolver.

**Changelog**
Update HEOM for v5.

hodgestar commented:
@Ericgig I've kept the WIP label, but this is ready for a first review. I think I'm made the HEOM look as much like the base solver as I can without having to modify the base solver a bit. Perhaps we can meet up tomorrow on Discord to discuss options.

I'm also happy to merge this as is and adapt the solver more in another PR (assuming tests still pass and documentation still builds with my latest commits).

Ericgig reviewed (commented):
As a fix to make HEOM works in `dev.major` and included in tests and docs, then it looks fine and I would not mind it being merge as is. 

But it is not ready for release. The way it tries to be a v5 solver is quite forceful. I don't think any modification to base solver is needed, but change the way it use it. A `HeomResult` to store the full ADO state could be useful. 
Also it seems a case that could from running on GPU, we should add better support data-layer.

I have doubt about having `steady_state` attached to solver: what about time dependent systems or what about `sesolve`.

hodgestar commented:
@Ericgig Apologies -- I forgot to commit my final changes from the weekend. :/

hodgestar commented on qutip/solve/nonmarkov/bofin_baths.py:
```diff
@@ -14,10 +14,16 @@
 from qutip.core.superoperator import spre, spost
 
 
+def _isequal(Q1, Q2, tol):
+    """ Return true if Q1 and Q2 are equal to within the given tolerance. """
+    return _data.iszero(_data.sub(Q1.data, Q2.data), tol=tol)
```
Agreed.

hodgestar replied:
I see this has been added to the v5 TODO list -- thanks!

hodgestar commented on qutip/solve/nonmarkov/bofin_solvers.py:
```diff
@@ -376,18 +364,18 @@ def extract(self, idx_or_label):
 
 
-class HEOMSolver:
+class HEOMSolver(Solver):
```
It isn't. I'd like to look at how hard it would be to share more by modifying Solver slightly -- e.g. maybe `.run()` could be reused if there were some hooks to allow adding additional data to the Result, and passing the ADOs to the e_ops.

hodgestar replied:
It is now inheriting all of `.run`, `.start` and `.step` from solver -- woot!

nwlambert reviewed (commented):
I ran some comparisons to the old version, and everything looks good.  There seems to be some speedup both to the RHS construction and the ODE solver for most cases.  

A couple of comments on the open questions:

1) Should HierarchyADOsState store ado_state as a Data object rather than a numpy array? 

Since each individual ADO is naturally a Qobj, having HierarchyADOsState store them an array or list of Qobjs makes sense  I guess, it would make the ado_state object a bit more transparent to the user (though  .extract() already does the conversion to a qobj for a requested ADO).

2) How should the ado_state be stored on a results object? In particular, how should the change in e_ops signature be handled (for the HEOM solver the e_ops callables also take the full ADO state).

One thought  on this, does it make sense to have the callable e_ops() take something like (t, rho, ado_state)? Then if someone just wants to do something functional with rho on its own they can, without needing to extract from  from ado_state?  Or is that breaking with some expected functionality in callable e_ops in the solvers?

3) Recheck the time dependent case.

From what I can see it looks good! Seems like basically you treat the entire RHS as an effective qobjevo?  I ran through various examples, and couldn't find any issues!

4) The HEOM can't really take new QObjEvo arguments each time run is called because calculating the RHS again is quite expensive. Maybe something can be done?

Right, I was gonna say I guess since the most expensive part is the _rhs() call in _calculate_rhs(), just reconstructing the diagonal time-dependent part on its own  might not be too bad, since that is what was done in the older version with the callback function. But I am not certain how all the moving parts of the solver stuff fit together now, so maybe there is a smart way to do it?

5) Compare performance to the HEOM in 4.7.

I didn't run lots of scaling examples or anything, but for standard small examples speed up was 4 or 5 times.  On the dynamical decoupling time dependent examples (which maybe have some other overhead in what I was doing) still seemed about 30% faster.  On the fermionic examples, with very-large system Hilbert space, there was a little slowdown in the RHS construction.

hodgestar commented:
> I ran some comparisons to the old version, and everything looks good. There seems to be some speedup both to the RHS construction and the ODE solver for most cases.

Thanks! I've now streamlined things a little more, so they might be slightly faster still.

> Since each individual ADO is naturally a Qobj, having HierarchyADOsState store them an array or list of Qobjs makes sense I guess

I've put this down in the v5 TODO list (just to draw a line under this PR).

>     2. How should the ado_state be stored on a results object? In particular, how should the change in e_ops signature be handled (for the HEOM solver the e_ops callables also take the full ADO state).

I ended up always calling `e_ops` functions with `f(t, ado_state)`. This is a bit different to other solvers, but makes what the HEOMSolver considers the evolution state consistent. The density matrix part of the state is available as `ado_state.rho` so it's not a big hardship to retrieve just the system state.

>>     3. Recheck the time dependent case.
> 
> From what I can see it looks good! Seems like basically you treat the entire RHS as an effective qobjevo? I ran through various > examples, and couldn't find any issues!

I simplified things a bit further. In the time-dependent case the RHS matrix is calculated without adding the Liouvillian to the diagonal and then the kron of the Liouvillian is added to the RHS at the end, resulting in a simpler QobjEvo.

>>    4. The HEOM can't really take new QObjEvo arguments each time run is called because calculating the RHS again is quite expensive. Maybe something can be done?

Arguments are supported now because the full RHS is a QobjEvo!

>>     5. Compare performance to the HEOM in 4.7.
> 
> On the fermionic examples, with very-large system Hilbert space, there was a little slowdown in the RHS construction.

Maybe the slowdown has gone away now.

hodgestar commented:
@nwlambert @Ericgig WIP removed -- ready for another review!

hodgestar commented:
@Ericgig Tests added for the type changes.

Note that  `QobjEvo.__rmatmul__` is not currently callable. Calling `qobj @ qevo` will call `QobjEvo.__matmul__(qobj, qevo)` first, which will return the result. If I am not missing a case where `__rmatmul__` is called, I vote that we simply delete the implementation of `__rmatmul__` (in a separate PR).

hodgestar commented:
> @Ericgig Tests added for the type changes.
> 
> Note that `QobjEvo.__rmatmul__` is not currently callable. Calling `qobj @ qevo` will call `QobjEvo.__matmul__(qobj, qevo)` first, which will return the result. If I am not missing a case where `__rmatmul__` is called, I vote that we simply delete the implementation of `__rmatmul__` (in a separate PR).

@Ericgig Explained that in Cython 3 `__rmatmul__` will be called, so we can keep it for now.

Ericgig reviewed (commented):
Overall, it look good as a way to make HEOM solvers work on v5.

But the use of data layer seems not ideal, why is the rhs fixed to `csr`. With it's size, it's probably not ideal as a Dense, but I don't think we should raise an error for other type... 

There is a few differences with other solvers:
  - There is no `heomsolve` to go with `HEOMSolver`.
  - With other solvers, they accept `H` as a `Qobj` or `QobjEvo` only, list format is accepted by the __solve function.

`bofin_*.py` does not have a `__all__`, even if they are not supposed to be imported directly, it would be better if they had.

The test coverage does not include time-dependent Hamiltonian. Are those too slow to include in automated tests?

hodgestar commented on qutip/solve/nonmarkov/bofin_solvers.py:
```diff
@@ -402,38 +421,38 @@ class HEOMSolver:
+    def __init__(self, H, bath, max_depth, *, options=None):
+        _time_start = time()
+
+        self.H_sys = QobjEvo(self._convert_h_sys(H))
```
Good idea! This was a left over from all the difficult refactoring when I originally moved BoFiN into QuTiP. I fixed the code to always store L_sys.

hodgestar commented:
> Overall, it look good as a way to make HEOM solvers work on v5.

Woot.

> But the use of data layer seems not ideal, why is the rhs fixed to `csr`. With it's size, it's probably not ideal as a Dense, but I don't think we should raise an error for other type...

It doesn't actually restrict the RHS to CSR -- it only *constructs* the RHS as a CSR initially. The base solver will convert the RHS to the correct format for the integrator if needed. It might be possible one day to avoid constructing the RHS as a CSR but it might be tricky to make it not horribly slow, so I'd like to leave this as potential enhancement rather than work on it right now.

> There is no `heomsolve` to go with `HEOMSolver`.

I've added `heomsolve`. It has more parameters than `sesolve` or `mesolve`, which is not great, but I'm not sure how to avoid it. The bath and the max_depth are not really optional for the HEOM, and there aren't reasonably default values.

> With other solvers, they accept `H` as a `Qobj` or `QobjEvo` only, list format is accepted by the __solve function.

I changed HEOMSolver to only support `Qobj` or `QobjEvo`. I allowed `HSolverDL` to continue to support other objects since `HSolverDL` exists only for backwards compatibility. Perhaps we should remove it at some point, but the amount of code in it is also tiny.

> `bofin_*.py` does not have a `__all__`, even if they are not supposed to be imported directly, it would be better if they had.

I've added `__all__`.
 
> The test coverage does not include time-dependent Hamiltonian. Are those too slow to include in automated tests?

The tests don't include a truly time-dependent case because it's hard to solve one analytically. Previously I tested the time-dependent logic by supplying `[H]` as the system (i.e. fooling the code into using the time-dependent path) but that broke when I switched to using `QobjEvo.isconstant`. I fixed it by using `[H, lambda t: 1.0]` instead, which would be very hard to determine as constant. Thank for picking this up.

Me might be able to do a truly time-dependent test by running with a memory-less bath and comparing with mesolve, but I'd also like to leave that as future work.

Ericgig reviewed (approved):
Look good.
I am thrusting you for the equations exactitude.

Heom could be moved from `qutip/solve` to `qutip/solver`, as the former is for v5 ready solver.

hodgestar commented:
Thanks @Ericgig. I'll move the HEOM into `qutip.solve` (and move tests and fix docs) in the morning and then merge if all the tests are happy, unless you want to have one final check?

Ericgig commented:
If it's just renaming files, I don't need a final check.
